### PR TITLE
HistogramData Rollout: LessThanMD to PlusMD

### LINKED ISSUE #18661: Histogram Data Rollout LessThanMD to PlusMD
Refactoring following algorithms to Histogram Data
- [x] Framework/MDAlgorithms/src/LessThanMD.cpp
- [x] Framework/MDAlgorithms/src/LoadILLAscii.cpp
- [x] Framework/MDAlgorithms/src/LoadSQW.cpp
- [x] Framework/MDAlgorithms/src/MDNormDirectSC.cpp
- [x] Framework/MDAlgorithms/src/MDNormSCD.cpp
- [x] Framework/MDAlgorithms/src/MinusMD.cpp
- [x] Framework/MDAlgorithms/src/MultiplyMD.cpp
- [x] Framework/MDAlgorithms/src/PlusMD.cpp

The below algorithms require additional work to implement serialisation for Histogram data types. That and the histogram data refactoring will be covered in issue #18717
- [ ] ~Framework/MPIAlgorithms/src/BroadcastWorkspace.cpp~
- [ ] ~Framework/MPIAlgorithms/src/GatherWorkspaces.cpp~

### PR DESCRIPTION
**Pull request by dtasev**
Description of work.

 Framework/MDAlgorithms/src/LessThanMD.cpp
 Framework/MDAlgorithms/src/LoadILLAscii.cpp
 Framework/MDAlgorithms/src/LoadSQW.cpp
 Framework/MDAlgorithms/src/MDNormDirectSC.cpp
 Framework/MDAlgorithms/src/MDNormSCD.cpp
 Framework/MDAlgorithms/src/MinusMD.cpp
 Framework/MDAlgorithms/src/MultiplyMD.cpp
 Framework/MDAlgorithms/src/PlusMD.cpp

The below algorithms require additional work to implement serialisation for Histogram data types. That and the histogram data refactoring will be covered in issue #18717
- [ ] ~Framework/MPIAlgorithms/src/BroadcastWorkspace.cpp~
- [ ] ~Framework/MPIAlgorithms/src/GatherWorkspaces.cpp~

**To test:**
- [ ] Run DataHandling tests
- [ ] Code inspection

Fixes #18661 .

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/MDAlgorithms/src/MultiplyMD.cpp [hunk 1]
```diff
@@ -47,8 +47,8 @@ void MultiplyMD::checkInputs() {
 template <typename MDE, size_t nd>
 void MultiplyMD::execEventScalar(typename MDEventWorkspace<MDE, nd>::sptr ws) {
   // Get the scalar multiplying
-  float scalar = float(m_rhs_scalar->dataY(0)[0]);
-  float scalarError = float(m_rhs_scalar->dataE(0)[0]);
+  float scalar = float(m_rhs_scalar->mutableY(0)[0]);
```
Comment (LamarMoore) [inline Framework/MDAlgorithms/src/MultiplyMD.cpp]: `auto scale = static_cast<float>(m_rhs_scalar->y(0)[0]);`?

File: Framework/MPIAlgorithms/src/BroadcastWorkspace.cpp [hunk 2]
```diff
@@ -91,9 +92,7 @@ void BroadcastWorkspace::exec() {
   for (std::size_t i = 0; i < numSpec; ++i) {
     if (world.rank() == root) {
       // For local output, just copy over
-      outputWorkspace->dataX(i) = inputWorkspace->readX(i);
-      outputWorkspace->dataY(i) = inputWorkspace->readY(i);
-      outputWorkspace->dataE(i) = inputWorkspace->readE(i);
+      outputWorkspace->setHistogram(i, inputWorkspace->histogram(i));
 
       // Send out the current spectrum
       broadcast(world, const_cast<MantidVec &>(inputWorkspace->readX(i)), root);
```
Comment (LamarMoore) [inline Framework/MPIAlgorithms/src/BroadcastWorkspace.cpp]: `readX`?

File: Framework/MPIAlgorithms/src/GatherWorkspaces.cpp [hunk 3]
```diff
@@ -202,9 +202,7 @@ void GatherWorkspaces::exec() {
                eplus(), 0);
       } else if (accum == "Append") {
         // Copy over data from own input workspace
-        outputWorkspace->dataX(wi) = inputWorkspace->readX(wi);
-        outputWorkspace->dataY(wi) = inputWorkspace->readY(wi);
-        outputWorkspace->dataE(wi) = inputWorkspace->readE(wi);
+        outputWorkspace->setHistogram(wi, inputWorkspace->histogram(wi));
 
         const int numReqs(3 * (included.size() - 1));
         std::vector<boost::mpi::request> reqs(numReqs);
```
Comment (LamarMoore) [inline Framework/MPIAlgorithms/src/GatherWorkspaces.cpp]: There are some `dataX` calls below this.

File: Framework/MPIAlgorithms/test/BroadcastWorkspaceTest.h [hunk 4]
```diff
@@ -57,10 +57,18 @@ class BroadcastWorkspaceTest : public CxxTest::TestSuite {
     API::MatrixWorkspace_const_sptr outWS =
         broadcaster.getProperty("OutputWorkspace");
     TS_ASSERT_EQUALS(inWS->size(), outWS->size());
+
+    const auto &inX = inWS->x(0);
```
Comment (LamarMoore) [inline Framework/MPIAlgorithms/test/BroadcastWorkspaceTest.h]: You didn't really need to do this. `inWS->x(0)[i]` would have been sufficient. You do not have to change it but it seems like a lot of unecessary extra code. Actually if you include `#include <MantidTestHelpers/HistogramDataTestHelpers.h`, header file name might not be correct, you can directly compare the two using `TS_ASSERT_EQUALS(inWS->x(0), outWS->x(0));`

File: Framework/MPIAlgorithms/test/GatherWorkspacesTest.h [hunk 5]
```diff
@@ -59,9 +59,9 @@ class GatherWorkspacesTest : public CxxTest::TestSuite {
         gatherer.getProperty("OutputWorkspace");
     TS_ASSERT_EQUALS(inWS->size(), outWS->size());
     for (int i = 0; i < 5; ++i) {
-      TS_ASSERT_EQUALS(inWS->readX(0)[i], outWS->readX(0)[i]);
-      TS_ASSERT_EQUALS(inWS->readY(0)[i], outWS->readY(0)[i]);
-      TS_ASSERT_EQUALS(inWS->readE(0)[i], outWS->readE(0)[i]);
+      TS_ASSERT_EQUALS(inWS->x(0)[i], outWS->x(0)[i]);
+      TS_ASSERT_EQUALS(inWS->y(0)[i], outWS->y(0)[i]);
+      TS_ASSERT_EQUALS(inWS->e(0)[i], outWS->e(0)[i]);
```
Comment (LamarMoore) [inline Framework/MPIAlgorithms/test/GatherWorkspacesTest.h]: See above comment.

File: Framework/MPIAlgorithms/test/GatherWorkspacesTest.h [hunk 6]
```diff
@@ -88,9 +88,9 @@ class GatherWorkspacesTest : public CxxTest::TestSuite {
...
```
Comment (LamarMoore) [inline Framework/MPIAlgorithms/test/GatherWorkspacesTest.h]: See above.

**Review by LamarMoore — CHANGES_REQUESTED** [review]
Looks good. I noticed in a few places you missed some of the legacy implementation. Make sure you check the files thoroughly. A simple search and find on the files might reveal something you have missed.

File: Framework/MPIAlgorithms/src/BroadcastWorkspace.cpp (hunk 2 above)
Comment (DimitarTasev) [inline Framework/MPIAlgorithms/src/BroadcastWorkspace.cpp]: Sorry, I didn't copy it from the issue description, BroadcastWorkspace and GatherWorkspaces need more work, as the `boost::mpi::broadcast(...)` function does not work with HistogramData types, and serialisation needs to be implemented. I've created an issue for that #18717, which will also cover the HistogramData refactoring

File: Framework/MPIAlgorithms/src/GatherWorkspaces.cpp (hunk 3 above)
Comment (DimitarTasev) [inline Framework/MPIAlgorithms/src/GatherWorkspaces.cpp]: Same comment as `BroadcastWorkspaces`

**Comment by DimitarTasev** [comment]
I've reverted any changes to `BroadcastWorkspace` and `GatherWorkspaces`, because I cannot build it locally, and I do not think that the build servers run the tests for MPIAlgorithms, so I cannot be sure that my changes do not break the scripts.

**Comment by LamarMoore** [comment]
@DimitarTasev can you make an issue for those to be refactored away from the legacy implementation?

**Comment by DimitarTasev** [comment]
@LamarMoore Issue is #18717 for both conversion and implementing the serialising

**Review by LamarMoore — APPROVED** [review]

**Comment by LamarMoore** [comment]
:shipit:

**Pull request merged**